feat: Add test documentation

JQrdan opened the pull request:
- Adds test overview documentation
- Adds documentation for test specific files and directories

Closes: #7

Commit:
feat: Add test documentation

- Adds test overview documentation
- Adds documentation for test specific files and directories

Signed-off-by: Jordan <jordan.tucker1@ibm.com>

dlabaj commented on docs/Test.md:
```diff
@@ -0,0 +1,150 @@
+  1. Perform any required setup
+  2. Perform an action (or actions)
+  3. Assert the result
+  4. Perform any required tidyup
```
can we change tidyup to clean up

dlabaj commented on docs/Test.md:
```diff
@@ -0,0 +1,150 @@
+
+## Style of test
+
+We have two styles of test - `Behavioural` and `Functional`.
```
Instead Behavioral style and Functional style do we just want to go with Behavior Drive Development and Test Driven Development?

JQrdan replied:
Fair point, I'll update these to be more familiar

dlabaj commented on docs/Test.md:
```diff
@@ -0,0 +1,150 @@
+- `client/Hooks`
+- `client/Contexts`
+
+### Behavioural test style
```
Where are we putting the behavioral and functional tests?  Are they going to live alongside the components, elements, hooks, contexts, and panel code in src?   Or are we going to put them somewhere else?

JQrdan replied:
They'll go alongside each component as a regular test would. The split between behavioural/functional is more of a guidance of what approach should be taken for each component. I'll add a piece that says this is where they should go

dlabaj reviewed (commented):
A few comments.

Commit:
refactor: address review comments

- Updated tidyup to clean up
- Behavioural/Functional now split into clearer BDD and TDD
- Added section to clarify where BDD and TDD tests should go

Signed-off-by: Jordan <jordan.tucker1@ibm.com>

matthew-chirgwin commented on docs/Test.md:
````diff
@@ -0,0 +1,154 @@
+
+## Test utilities
+
+Where possible, single responsibility test utilities should be created
+to abstract common test logic so that it can be easily re-used and lifts
+the barrier to entry to write new tests. These utilities should have simple
+function signatures that are well documented. Once implemented these should
+not be changed so that function complexity does not grow. If additional behaviour
+is required later on, these should be extended through new utility functions.
+
+## Testing contexts (an example of using test utilities)
+
+Testing contexts is an example of having common test logic that can be
+abstracted away. Here, the function `renderWithContextProviders` has been
+provided that extends the base rtl `render` function but can be given a list
+of context providers and their values that will wrap the provided children.
+Using this, a common render can be defined in a test file for contexts so that
+each test only needs to worry about using a consumer. For example:
+
+```
+render = (children) =>
+  renderWithContextProviders(children, {}, [
+    { provider: MyProvicer1, value: providerValue1 },
+    { provider: MyProvider2, value: providerValue2 },
+  ]);
+```
+
+This can be defined in a test file so that tests can still use a `render`
+function in the same way they usually would with rtl but these renders will be
+wrapped in the providers defined in the above function.
````
Would suggest linking off to the `utils/test/README` here, and having links in that README to specific helpers (like the one here for contexts). I can see this list growing, making this document harder to follow.

matthew-chirgwin commented on client/Bootstrap/README.md:
```diff
@@ -6,7 +6,7 @@ While it's own directory, the code in this directory should be tested and develo
 ## Test approach
 
-To be defined in https://github.com/strimzi/strimzi-ui/issues/7
+Bootstrap should be tested in a behavioural manor. See [Behavioural Driven Development](../../docs/Test.md#style-of-test).
```
I think it would be worth adding here that the tests for `Bootstrap` level components (and the whole UI more generally) are kept in the `e2e` folder (with a link to it).

JQrdan replied:
Happy to say bootstrap needs to only be tested through E2E as that is much more meaningful than trying to make fiddly unit tests. Hesitant about the navigation pieces being E2E only but willing to at least try, depending on its implementation. @dlabaj do you have any thoughts on this?

matthew-chirgwin commented on e2e/README.md:
```diff
@@ -0,0 +1,3 @@
+# End to End testing
+
+Please find documentation for E2E testing [here](../../docs/Test.md#end-to-end-testing).
```
Would suggest having a brief summary here of what this directory contains (along with links to sub folder readmes)

matthew-chirgwin commented on docs/Test.md:
```diff
@@ -0,0 +1,154 @@
+
+## Code coverage
+
+In this repo, we aim to have 100% code coverage across all components. This code
```
I had an idea around coverage: based on the files changed, run just related tests via Jest (for example), and get coverage on just those results. Ie, tests related to the change made to check coverage - no chance of having another test happen to 'run over' lines.

I can see this being non trivial, but will mean an area can be fully excercised by it's tests alone.

JQrdan replied:
For now, I'll just expand on what we believe the use of code coverage is. Further investigation on selective test coverage is needed before this can be fleshed out. I have raised an investigation point on the project board https://github.com/orgs/strimzi/projects/2#card-46309994.

matthew-chirgwin commented on utils/test/README.md:
```diff
@@ -0,0 +1,3 @@
+# Test Utils
+
+This directory contains all files that provide utility to all tests. See [Test utilities](../../docs/Test.md#test-utilities) .
```
See https://github.com/matthew-chirgwin/strimzi-ui/tree/master/utils/build for example of how this file could look

matthew-chirgwin reviewed (changes requested):
Looks good! Other than comments already left, just a few things which would suggest we add:

- Add a link to `docs/README.md` to the Test doc
- Add a link on the main repo README to the test document

Commit:
refactor: updates from review comments

Signed-off-by: Jordan <jordan.tucker1@ibm.com>

matthew-chirgwin commented on docs/Test.md:
```diff
@@ -0,0 +1,143 @@
+
+### Server Testing
+
+The strimzi-ui server should be tested from outside the server - like how a
```
`The strimzi-ui server should be tested from outside the server - like how a user would use it`. I think this might need a slight grammar tweak :)  Honestly, I am not sure if this sentence adds much. The main thing to know is the server code is/should be tested in a behavioural manner, as a user would acessing it (indirectly) via the UI. I think the second sentence (with a few extra words) could do the job.

matthew-chirgwin reviewed (commented):
LGTM, bar a question around grammar - thanks :)

Commit:
refactor: update server test section

Signed-off-by: Jordan <jordan.tucker1@ibm.com>

matthew-chirgwin commented:
Given there have been no further comments, merging this now.
